### ISSUE TITLE
BUG: Hashing/caching in parallel

### ISSUE THREAD

**Issue by will-saunders-ukaea** [firedrakeproject/firedrake#3732]
**Describe the bug**
Running with multiple ranks intermittently causes errors to be thrown like:
```
Traceback (most recent call last):
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/interpolation.py", line 837, in _interpolate
    assembled_interpolator = self.frozen_assembled_interpolator
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'SameMeshInterpolator' object has no attribute 'frozen_assembled_interpolator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/global_kernel.py", line 349, in __call__
    func = self._func_cache[key]
           ~~~~~~~~~~~~~~~~^^^^^
KeyError: 136635048558704

During handling of the above exception, another exception occurred:
```

I have a Firedrake install from 17/07/2024 which does not exhibit the issue and an install from 12/08/2024 which does.

**Steps to Reproduce**
WIP MFE - As the error is intermittent getting an backtrace to isolate the behaviour to make a MFE around is work in progress.

I have had the most success reproducing the error by:
1) cleaning the cache.
2) launch code sequentially (runs)
3) launch with 2 ranks

Working on MFE.

**Expected behaviour**
Runs without error.

**Error message**
* See above.

**Environment:**
 - OS: Ubuntu 22.04
 - Python version: 3.11.9
 - Output of `firedrake-status`
 
```
firedrake-status
/home/js0259/venvs/nesodrake-2024_08_12/bin/firedrake-status:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  __import__('pkg_resources').require('firedrake==0.13.0+6164.g57e21cc8e')
Firedrake Configuration:
    package_manager: True
    minimal_petsc: False
    mpicc: /opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/mpich-4.2.1-l4h3dm4laltqu6p2jeeaathwobbulxpf/bin/mpicc
    mpicxx: /opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/mpich-4.2.1-l4h3dm4laltqu6p2jeeaathwobbulxpf/bin/mpicxx
    mpif90: /opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/mpich-4.2.1-l4h3dm4laltqu6p2jeeaathwobbulxpf/bin/mpif90
    mpiexec: /opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/mpich-4.2.1-l4h3dm4laltqu6p2jeeaathwobbulxpf/bin/mpiexec
    disable_ssh: True
    honour_petsc_dir: False
    with_parmetis: False
    slepc: False
    packages: ['git+ssh://github.com/firedrakeproject/Irksome.git#egg=Irksome']
    honour_pythonpath: True
    opencascade: False
    torch: False
    petsc_int_type: int32
    cache_dir: /home/js0259/venvs/nesodrake-2024_08_12/.cache
    complex: False
    remove_build_files: False
    with_blas: None
    netgen: True
Additions:
    None
Environment:
    PYTHONPATH: /opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/py-setuptools-69.2.0-gdrglctx2y4mavk5uxwkjmcbklf6tbnt/lib/python3.11/site-packages:/opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/py-pip-23.1.2-h6dctfqbarcodybop2xrnel5re3bod4b/lib/python3.11/site-packages:/opt/spack/v0.22-gcc12/linux-ubuntu22.04-cascadelake/gcc-12.3.0/python-venv-1.0-fdfirut25kl53qktsowvq4nniij6mjoc/lib/python3.11/site-packages
    PETSC_ARCH: None
    PETSC_DIR: None
Status of components:
---------------------------------------------------------------------------
|Package             |Branch                        |Revision  |Modified  |
---------------------------------------------------------------------------
|FInAT               |master                        |eaee8fe   |False     |
|Irksome             |master                        |4526f43   |False     |
|PyOP2               |master                        |5f18075f  |False     |
|fiat                |master                        |6b53feb   |False     |
|firedrake           |master                        |57e21cc8e |False     |
|h5py                |firedrake                     |db2ab02f  |False     |
|libsupermesh        |master                        |c07de13   |False     |
|loopy               |main                          |d9876d83  |False     |
|ngsPETSc            |main                          |7afc820   |False     |
|petsc               |firedrake                     |3a8a2250205|False     |
|pyadjoint           |master                        |92121af   |False     |
|pytest-mpi          |main                          |f2566a1   |False     |
|tsfc                |master                        |8db60e0   |False     |
|ufl                 |master                        |c1a8afb1  |False     |
---------------------------------------------------------------------------
```

**Comment by will-saunders-ukaea** [comment]
Another backtrace example:
```
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/interpolation.py", line 837, in _interpolate
    assembled_interpolator = self.frozen_assembled_interpolator
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'SameMeshInterpolator' object has no attribute 'frozen_assembled_interpolator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/global_kernel.py", line 349, in __call__
    func = self._func_cache[key]
           ~~~~~~~~~~~~~~~~^^^^^
KeyError: 125673324463536

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/js0259/git-ukaea/NESODrake-TwoStream/examples/prototype.py", line 164, in <module>
    out_rho.write(rho, poisson_rhs)
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/output/vtk_output.py", line 648, in write
    vtu = self._write_vtu(*functions)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/output/vtk_output.py", line 531, in _write_vtu
    functions = tuple(self._prepare_output(f, max_elem)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/output/vtk_output.py", line 531, in <genexpr>
    functions = tuple(self._prepare_output(f, max_elem)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/pyadjoint/pyadjoint/tape.py", line 110, in wrapper
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/output/vtk_output.py", line 486, in _prepare_output
    output.interpolate(function)
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/function.py", line 413, in interpolate
    return assemble(interp, tensor=self, ad_block_tag=ad_block_tag)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/adjoint_utils/assembly.py", line 30, in wrapper
    output = assemble(form, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/assemble.py", line 133, in assemble
    return get_assembler(expr, *args, **kwargs).assemble(tensor=tensor)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/assemble.py", line 381, in assemble
    result = BaseFormAssembler.base_form_postorder_traversal(self._form, visitor, visited)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/assemble.py", line 603, in base_form_postorder_traversal
    visited[e] = visitor(e, *(visited[arg] for arg in operands))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/assemble.py", line 377, in visitor
    return self.base_form_assembly_visitor(e, t, *operands)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/assemble.py", line 550, in base_form_assembly_visitor
    return firedrake.Interpolator(expression, tensor, **interp_data)._interpolate(default_missing_val=default_missing_val)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/interpolation.py", line 840, in _interpolate
    assembled_interpolator = self.callable()
                             ^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/firedrake/firedrake/interpolation.py", line 1005, in callable
    l()
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/parloop.py", line 241, in compute
    self()
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/parloop.py", line 251, in __call__
    self._compute(self.iterset.core_part)
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/parloop.py", line 232, in _compute
    self.global_kernel(self.comm, part.offset, part.offset+part.size, *self.arglist)
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/global_kernel.py", line 351, in __call__
    func = self.compile(comm)
           ^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/global_kernel.py", line 422, in compile
    return compilation.load(self, extension, self.name,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/compilation.py", line 596, in load
    dll = compiler(cppargs, ldargs, cpp=cpp, comm=comm).get_so(code, extension)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/js0259/venvs/nesodrake-2024_08_12/src/PyOP2/pyop2/compilation.py", line 339, in get_so
    raise CompilationError("Generated code differs across ranks (see output in %s)" % output)
pyop2.exceptions.CompilationError: Generated code differs across ranks (see output in /home/js0259/venvs/nesodrake-2024_08_12/.cache/pyop2/mismatching-kernels)
Abort(1) on node 1 (rank 1 in comm 496): application called MPI_Abort(PYOP2_COMM_WORLD, 1) - process 1
```

**Comment by connorjward** [comment]
Can you share the differing bits of generated code?

**Comment by JDBetteridge** [comment]
This looks suspiciously similar to a parallel cache coherency issue I ran into recently (noted in [these minutes](https://github.com/firedrakeproject/firedrake/wiki/Firedrake-meeting-2024-08-07#memorydisk-caches-are-broken-in-parallel)). Are you running your code over multiple communicators, or just `COMM_WORLD`?

Also it would be good to see the differing code as Connor says: Take a look in `/home/js0259/venvs/nesodrake-2024_08_12/.cache/pyop2/mismatching-kernels`

**Comment by will-saunders-ukaea** [comment]
On the Firedrake side there is just COMM_WORLD. We have some communicators that we create/destroy in C++, these are never passed to Firedrake,

**Comment by will-saunders-ukaea** [comment]
At the interpolation call [11] one rank is trying to interpolate a scalar valued function and the other rank a vector valued function. I do have  vtu output calls for a vector valued function and a scalar valued function, both the backtraces are on the scalar valued write call.

```
# One rank:

(Pdb++) p expression
Coefficient(WithGeometry(FunctionSpace(<firedrake.mesh.MeshTopology object at 0x728e9c1041d0>, VectorElement(FiniteElement('Discontinuous Lagrange', triangle, 1, variant='equispaced'), dim=2, variant='equispaced'), name=None), Mesh(VectorElement(FiniteElement('Discontinuous Lagrange', triangle, 1, variant='equispaced'), dim=2, variant='equispaced'), 14)), 53)
(Pdb++)

# The other rank:

(Pdb++) p expression
Coefficient(WithGeometry(FunctionSpace(<firedrake.mesh.MeshTopology object at 0x734d44054410>, FiniteElement('Discontinuous Lagrange', triangle, 1), name=None), Mesh(VectorElement(FiniteElement('Discontinuous Lagrange', triangle, 1, variant='equispaced'), dim=2, variant='equispaced'), 14)), 36)
```

**Comment by will-saunders-ukaea** [comment]
One rank, the "vector" one, is on line 529 of `firedrake/output/vtk_output.py` processing the mesh coordinates and the second rank, the "scalar" one, has progressed to line 531 to process the actual function to write.

**Comment by will-saunders-ukaea** [comment]
I haven't seen it fail if I comment out the solve call.

**Comment by colinjcotter** [comment]
not the issue here but you can use PETSc.Sys.Print to ensure only one print statement in parallel.

**Comment by will-saunders-ukaea** [comment]
> not the issue here but you can use PETSc.Sys.Print to ensure only one print statement in parallel.

Oh nice, thanks.

**Comment by JDBetteridge** [comment]
I had hoped that this [FInAT PR](https://github.com/FInAT/FInAT/pull/134) and this [Firedrake PR](https://github.com/firedrakeproject/firedrake/pull/3730) would have sorted the issue. Unfortunately not!

It seems that [the error is here](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/output/vtk_output.py#L462) on line 462.

A `finat.finiteelementbase.FiniteElementBase` is being used for comparison, but it does not have a stable (or even useful) hash as noted in [this comment](https://github.com/FInAT/FInAT/pull/134#discussion_r1736358798). I think one solution would be to use a `finat.ufl.finiteelementbase.FiniteElementBase` subclass here. Maybe someone knows better than me how we fix this!

**Comment by connorjward** [comment]
> I had hoped that this [FInAT PR](https://github.com/FInAT/FInAT/pull/134) and this [Firedrake PR](https://github.com/firedrakeproject/firedrake/pull/3730) would have sorted the issue. Unfortunately not!
> 
> It seems that [the error is here](https://github.com/firedrakeproject/firedrake/blob/master/firedrake/output/vtk_output.py#L462) on line 462.
> 
> A `finat.finiteelementbase.FiniteElementBase` is being used for comparison, but it does not have a stable (or even useful) hash as noted in [this comment](https://github.com/FInAT/FInAT/pull/134#discussion_r1736358798). I think one solution would be to use a `finat.ufl.finiteelementbase.FiniteElementBase` subclass here. Maybe someone knows better than me how we fix this!

I think for this case we usually compare UFL elements instead of FInAT ones. We just should be careful because sometimes we don't consider vector-ness when we make the FInAT element.

I think we should probably use the UFL element here and perhaps make __eq__ and/or __hash__ raise an error for FInAT elements to catch any more of these.

**Comment by connorjward** [comment]
@will-saunders-ukaea can you try `connorjward/fix-3732` (#3807) and see if that resolves things for you? It did for me.

Then just need to ensure that the remaining tests pass.

**Comment by JDBetteridge** [comment]
Seems to be fixed, closing
